package/resource_monitor/resource_monitor/src/query_process_file_descriptor.c

```diff
@@ -0,0 +1,499 @@
+        free(file_entry);
+        return false;
+      }
+    }
+    file_entry->fd_count++;
+  }
+  return true;
+}
+
+static bool parse_fd_line(resq_state_t *state, const char *line)
+{
+  char id_str[16] = {0};
+  char exe_str[64] = {0};
+  char file_str[128] = {0};
+  line_spec_t line_specs[FD_STATE_COUNT] = {[FD_STATE_ID_STR] =
+                                              (line_spec_t){
+                                                .type = FT_STR,
+                                                .dst.str = id_str,
+                                                .buflen = sizeof(id_str),
+                                                .desc = "pid",
+                                                .next = FD_STATE_EXE_FILE_STR,
+                                              },
+                                            [FD_STATE_EXE_FILE_STR] =
+                                              (line_spec_t){
+                                                .type = FT_STR,
+                                                .dst.str = exe_str,
+                                                .buflen = sizeof(exe_str),
+                                                .desc = "executable",
+                                                .next = FD_STATE_FD_STR,
+                                              },
+                                            [FD_STATE_FD_STR] =
+                                              (line_spec_t){
+                                                .type = FT_STR,
+                                                .dst.str = file_str,
+                                                .buflen = sizeof(file_str),
+                                                .desc = "file descriptor",
+                                                .next = FD_STATE_DONE,
+                                              },
+                                            [FD_STATE_DONE] = (line_spec_t){
+                                              .desc = "done",
+                                            }};
+
+  if (!parse_tab_line(line, FD_STATE_ID_STR, FD_STATE_DONE, line_specs, "\t")) {
+#ifdef DEBUG_QUERY_FD_TAB
+    PK_LOG_ANNO(LOG_ERR | LOG_SBP, "file descriptor query: parse_tab_line failed ");
+#endif
+    return false;
+  }
+  return process_fd_entry(state, file_str, id_str);
+}
+
+static void destroy_table_entry(table_t *table, void **entry)
+{
+  (void)table;
+  free(*entry);
+  *entry = NULL;
+}
+
+static int compare_counts(const void *ptc1, const void *ptc2)
+{
+  const id_to_count_t *id_to_count1 = ptc1;
+  const id_to_count_t *id_to_count2 = ptc2;
+  return -((int)id_to_count1->count - (int)id_to_count2->count);
+}
+
+static void run_resource_query(void *context)
+{
+  resq_state_t *state = context;
+  if (state->pid_table != NULL) {
+    table_destroy(&state->pid_table);
+  }
+  if (state->file_table != NULL) {
+    table_destroy(&state->file_table);
+  }
+  char leftover_buffer[4096] = {0};
```

Comment:
Since a buffer can potentially have no newlines in it, the `leftover_buffer` should be at least as big as the read buffer, suggest using `LSOF_CMD_BUF_LENGTH` to size this buffer.